### PR TITLE
Fix for DF training

### PR DESCRIPTION
**Pull request by masdevas**
Changes proposed in this pull request:
- Merge redundant memory allocations into existing allocations (and repair memory leaks)
- Fill undefined values "featureValueOrResponse" and "leftIndexOrClass" in leaves for classification and regression

### CODE REVIEW

**Comment by masdevas** [comment]
CI: http://intel-ci.intel.com/ea4b8249-be95-f1a9-b8dd-8cdcd4b723a1

**Comment by ShvetsKS** [comment]
Please provide some evidence with fixed fail in weekly testing.

File: algorithms/kernel/dtrees/dtrees_model_impl_common.h [hunk 1]
```diff
@@ -68,6 +68,7 @@ void nodeToTable(const NodeBase& node, size_t iRow, size_t& iCur, DecisionTreeNo
             }
         }
         row.featureIndex = -1;
+        row.featureValueOrResponse = 0;
```
Comment (ShvetsKS) [inline algorithms/kernel/dtrees/dtrees_model_impl_common.h]: in regression case we have:
void writeLeaf(const TLeaf& l, DecisionTreeNode& row)
{
    row.featureValueOrResponse = l.response;
}
This field `featureValueOrResponse` is not  initialized only for classification flow.
Seems we have to add the same for regression flow: `row.leftIndexOrClass = 0`

**Comment by ShvetsKS** [comment]
Did you plane to fix failed test cases in weekly scope?
Is it fix will be covered in test system?

**Pull request merged**